### PR TITLE
commitlog: Handle oversized entries (partially)

### PR DESCRIPTION
**Pull request by elcallio**
Refs #18161

Yet another approach to dealing with large commitlog submissions. This approach does not change any file format or file distribution. It only handles cases where the oversized write is actually a multi- write of N entries, where each sub-entry does in fact fit into a segment.

This does not change any rules on size limits.
However, oversized multi-allocs are automatically accepted if we can fit them using new tactics. This might break some dtest(?)

If an allocation is to large, we attempt a slow path, where we effectively lock out all other writers, and write each entry in turn, potentially to different chunks/segments.

To ensure "atomicity" we both force syncronization, and should the whole op fail, we restore segment state (rewinding), thus discarding data all we wrote.

Replaying will pick up entries in order, though, same as "normal" path, entries could be lost due to file corruption.

### CODE REVIEW

**Comment by avikivity** [comment]
I don't see documentation changes. Presumably the format has to be extended for multi-segment entries?

**Comment by elcallio** [comment]
No, this changes nothing about the format. It just splits the multi-entries into individual ones, but uses lock-out + possible rewind to guarantee all-or-nothing write of the full entry set.

**Comment by nyh** [comment]
In addition to a unit test which will (I hope?) exercises the edge cases like replay and atomicity, I think you should also add a functional test, e.g., in the test/cql-pytest framework or in whatever other framework you prefer. 
I.e., if I understood correctly, after this patch you are able to write a large BATCH (?) that wasn't possible before this patch. Wouldn't be nice to show it in a CQL example - a CQL command that works with the patch and failed previously?
Especially (and with cql-pytest you can actually test this) this limitation didn't happen in Cassandra but did in Scylla.

**Comment by elcallio** [comment]
The idea is that a few of the topology tests might be able to reach new, existing heights by allowing schema changes to be larger... So I think there are macro-tests, but sure, there is probably place for more. 
I'd prefer to get @avikivity :s view on this approach first though.

**Comment by avikivity** [comment]
What are the constraints of the solution? What are multi-entries?

For example, if the limitation is that an individual table write is limited, but writes to multiple tables aren't, then it won't work since I expect there will usually be a dominant table.

If a single table commit can be split into multiple sub-entries, then it works, but it punts the problem of how to split entries upwards. How can the caller make this determination?

**Comment by avikivity** [comment]
> To ensure "atomicity" we both force syncronization, and should the whole op fail, we restore segment state (rewinding), thus discarding data all we wrote.

What if we crash before rewinding?

**Comment by elcallio** [comment]
> What are the constraints of the solution? What are multi-entries?

See https://github.com/scylladb/scylladb/blob/f62197ee1efe58195d7f2b3e05afbe29470d726c/replica/database.cc#L1909

Used for applying N mutations in one go. It does not split a mutation, but atm, the cases where we fail due to over large schema changes (afaik) is due to such  N-mutation writes, not single, large mutations. Tablet project has worked trying to make topology/schema changes break up into smaller, distinct mutations. However, for lwt/raft we want to apply all these "atomically" together.

Splitting a mutation into more than one chunk/file would require changing the whole format, as well as all mechanisms regarding replay and more importantly ref-counting from memtable to segments etc. 

> What if we crash before rewinding?

Then we might have one or more mutations in the log, and some not. This is however no difference from the current state, where corruption/crash writing when one sector but not next might cause the same to happen.

**Comment by avikivity** [comment]
> > What are the constraints of the solution? What are multi-entries?
> 
> See
> 
> https://github.com/scylladb/scylladb/blob/f62197ee1efe58195d7f2b3e05afbe29470d726c/replica/database.cc#L1909
> 
> Used for applying N mutations in one go. It does not split a mutation, but atm, the cases where we fail due to over large schema changes (afaik) is due to such N-mutation writes, not single, large mutations. Tablet project has worked trying to make topology/schema changes break up into smaller, distinct mutations. However, for lwt/raft we want to apply all these "atomically" together.


I'm worried that we have cases where we do have a large single mutation:
 - fetching a raft snapshot
 - writing a system.tablets mutation after a large split

I don't want different fixes for different cases of the same underlying problem.

> 
> Splitting a mutation into more than one chunk/file would require changing the whole format, as well as all mechanisms regarding replay and more importantly ref-counting from memtable to segments etc.
> 
> > What if we crash before rewinding?
> 
> Then we might have one or more mutations in the log, and some not. This is however no difference from the current state, where corruption/crash writing when one sector but not next might cause the same to happen.

Right now we don't violate atomicity. If a mutation is split across several pages, then and not all pages recover, then we don't apply it. Two different mutations can be applied individually, but they are part of two separate commits.

btw for Raft we have to return failures if we fail to apply any mutation that isn't the last one, since it indicates we're missing data.

**Comment by elcallio** [comment]
> I don't want different fixes for different cases of the same underlying problem.

It is not the same problem. One is the problem of huge mutations. The second is the problem of applying N mutations and that we treat it like a single, large mutation. 

> Right now we don't violate atomicity. If a mutation is split across several pages, then and not all pages recover, then we don't apply it. Two different mutations can be applied individually, but they are part of two separate commits.

Which is the same as with this PR. Again, this is not one mutation, but the applying of a number of them. Today, even though they are written as a single large block, in the file they still exist, and are read, as a sequence of independent mutations, and can be lost/applied individually. This patch changes nothing about this. It (again) only allows for splitting a
N-mutation array into several chunks, but still treat it semi-atomic from the viewpoint of the writer.

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -323,6 +323,8 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     requires std::derived_from<T, db::commitlog::entry_writer> && std::same_as<R, decltype(std::declval<T>().result())>
     future<R> allocate_when_possible(T writer, db::timeout_clock::time_point timeout);
 
+    future<> oversized_allocation(entry_writer&, db::timeout_clock::time_point, size_t);
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:326]: nit: in commit message s/to large/too large/

**Comment by avikivity** [comment]
> > I don't want different fixes for different cases of the same underlying problem.
> 
> It is not the same problem. One is the problem of huge mutations. The second is the problem of applying N mutations and that we treat it like a single, large mutation.


I don't understand. Aren't they both a single add_entry() call?

> 
> > Right now we don't violate atomicity. If a mutation is split across several pages, then and not all pages recover, then we don't apply it. Two different mutations can be applied individually, but they are part of two separate commits.
> 
> Which is the same as with this PR. Again, this is not one mutation, but the applying of a number of them. Today, even though they are written as a single large block, in the file they still exist, and are read, as a sequence of independent mutations, and can be lost/applied individually. This patch changes nothing about this. It (again) only allows for splitting a N-mutation array into several chunks, but still treat it semi-atomic from the viewpoint of the writer.

I see. 

But shouldn't we provide atomicity?

I guess Raft doesn't really need it. The Raft log entry will be a single mutation. It will then be decomposed to individual mutations and re-committed. But it will remain in the Raft log until it's fully applied.

@kbr-scylla @gleb-cloudius comments?

**Comment by elcallio** [comment]
> I don't understand. Aren't they both a single add_entry() call?
> 
The call in this case is `add_entries`, which, as the name implies, is adding a vector of mutation entries, but treating the adding as a single, "atomic" (for some definition thereof) operation, unlike repeated calls to `add_entry`. 

> But shouldn't we provide atomicity?

Well, we do, to an extent (like being a little pregnant) - the apply is atomic, but if we crash in the middle, or file corruption causes parts of the segments to be lost, we might not be able to fully recover all parts of the data written. There we are again back to the question of "should a node that senses actual corruption of CL even start?" question.

**Comment by bhalevy** [comment]
> There we are again back to the question of "should a node that senses actual corruption of CL even start?" question.

We may have different answers for the system commitlog, which raft depend on, where we would want to not start if there are indications of corruption, vs. the user (default) commitlog, where running repair is possible to bring the node (and potentially views/indices) up to speed.
Ideally (and with tablets), that repair could be triggered automatically after the node starts.

**Comment by mykaul** [comment]
Can we get a design document before implementation? I don't mind POC and arguing over real code, but it has to be accompanied by documentation.

**Comment by tgrabiec** [comment]
> > I don't understand. Aren't they both a single add_entry() call?
> 
> The call in this case is `add_entries`, which, as the name implies, is adding a vector of mutation entries, but treating the adding as a single, "atomic" (for some definition thereof) operation, unlike repeated calls to `add_entry`.
> 
> > But shouldn't we provide atomicity?
> 
> Well, we do, to an extent (like being a little pregnant) - the apply is atomic, but if we crash in the middle, or file corruption causes parts of the segments to be lost, we might not be able to fully recover all parts of the data written. There we are again back to the question of "should a node that senses actual corruption of CL even start?" question.

I think a process crash, anywhere, shouldn't result in commitlog to be corrupted or considered as corrupted. Crashing should be safe, since we do it a lot.

**Comment by avikivity** [comment]
> > > I don't understand. Aren't they both a single add_entry() call?
> > 
> > 
> > The call in this case is `add_entries`, which, as the name implies, is adding a vector of mutation entries, but treating the adding as a single, "atomic" (for some definition thereof) operation, unlike repeated calls to `add_entry`.
> > > But shouldn't we provide atomicity?
> > 
> > 
> > Well, we do, to an extent (like being a little pregnant) - the apply is atomic, but if we crash in the middle, or file corruption causes parts of the segments to be lost, we might not be able to fully recover all parts of the data written. There we are again back to the question of "should a node that senses actual corruption of CL even start?" question.
> 
> I think a process crash, anywhere, shouldn't result in commitlog to be corrupted or considered as corrupted. Crashing should be safe, since we do it a lot.

Exactly. In fact we should crash more, in testing, and verify that commitlog is never corrupted, and verify that if we corrupt commitlog on purpose we detect it.

**Comment by elcallio** [comment]
Closing in favour of #19472
